File: lambda-functions/src/lambda_functions/github_store.py
import base64
import json
import os
import shutil
import tarfile
import tempfile
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Any


CONTENT_IMAGE_EXTENSIONS = {".png", ".jpg", ".jpeg", ".gif", ".webp", ".svg"}


class GitHubError(RuntimeError):
    pass


class GitHubStore:
    def __init__(self, root: Path):
        self.root = root
        self.content_root = root / "content"
        self.owner = os.environ.get("GITHUB_OWNER", "DataTalksClub")
        self.repo = os.environ.get("GITHUB_REPO", "dataops")
        self.branch = os.environ.get("GITHUB_BRANCH", "main")
        self.token = github_token()
        self._tree: dict[str, dict[str, Any]] | None = None
        self._synced = False

    @property
    def github_url(self) -> str:
        return f"https://github.com/{self.owner}/{self.repo}"

    def reset(self) -> None:
        self._tree = None
        self._synced = False
        if self.root.exists():
            shutil.rmtree(self.root)

    def sync_markdown(self) -> None:
        if self._synced:
            return
        self.root.mkdir(parents=True, exist_ok=True)
        self.content_root.mkdir(parents=True, exist_ok=True)
        self.download_markdown_tarball()
        self._synced = True

    def download_markdown_tarball(self) -> None:
        url = f"https://api.github.com/repos/{self.owner}/{self.repo}/tarball/{quote_path(self.branch)}"
        headers = {
            "accept": "application/vnd.github+json",
            "authorization": f"Bearer {self.token}",
            "user-agent": "dataops-lambda",
            "x-github-api-version": "2022-11-28",
        }
        req = urllib.request.Request(url, headers=headers, method="GET")
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                with tempfile.NamedTemporaryFile(dir="/tmp", suffix=".tar.gz", delete=False) as tmp:
                    shutil.copyfileobj(resp, tmp)
                    tmp_path = Path(tmp.name)
        except urllib.error.HTTPError as exc:
            detail = exc.read().decode("utf-8", "replace")
            raise GitHubError(f"GitHub tarball download failed: HTTP {exc.code}: {detail}") from exc

        try:
            with tarfile.open(tmp_path, "r:gz") as archive:
                for member in archive:
                    if not member.isfile():
                        continue
                    parts = member.name.split("/", 1)
                    if len(parts) != 2:
                        continue
                    repo_path = parts[1]
                    if not should_hydrate_tarball_path(repo_path):
                        continue
                    extracted = archive.extractfile(member)
                    if extracted is None:
                        continue
                    self._write_repo_file(repo_path, extracted.read())
        finally:
            tmp_path.unlink(missing_ok=True)

    def ensure_file(self, repo_path: str) -> Path:
        repo_path = normalize_repo_path(repo_path)
        local = self.local_path(repo_path)
        if local.exists():
            return local
        entry = self.tree().get(repo_path)
        if not entry or entry.get("type") != "blob":
            raise FileNotFoundError(repo_path)
        self._write_repo_file(repo_path, self.blob_bytes(entry["sha"]))
        return local

    def local_path(self, repo_path: str) -> Path:
        repo_path = normalize_repo_path(repo_path)
        target = (self.root / repo_path).resolve()
        if self.root not in target.parents and target != self.root:
            raise ValueError("Path escapes GitHub cache root")
        return target

    def put_local_file(self, repo_path: str, message: str) -> None:
        repo_path = normalize_repo_path(repo_path)
        local = self.local_path(repo_path)
        content = local.read_bytes()
        current_sha = self.current_sha(repo_path)
        body: dict[str, Any] = {
            "message": message,
            "content": base64.b64encode(content).decode("ascii"),
            "branch": self.branch,
        }
        if current_sha:
            body["sha"] = current_sha
        self.request("PUT", f"/repos/{self.owner}/{self.repo}/contents/{quote_path(repo_path)}", body)
        self.refresh_tree()

    def delete_repo_file(self, repo_path: str, message: str) -> None:
        repo_path = normalize_repo_path(repo_path)
        current_sha = self.current_sha(repo_path)
        if not current_sha:
            return
        body = {"message": message, "sha": current_sha, "branch": self.branch}
        self.request("DELETE", f"/repos/{self.owner}/{self.repo}/contents/{quote_path(repo_path)}", body)
        self.refresh_tree()

    def current_sha(self, repo_path: str) -> str:
        entry = self.tree().get(normalize_repo_path(repo_path))
        if not entry or entry.get("type") != "blob":
            return ""
        return str(entry.get("sha") or "")

    def tree(self) -> dict[str, dict[str, Any]]:
        if self._tree is None:
            data = self.request("GET", f"/repos/{self.owner}/{self.repo}/git/trees/{quote_path(self.branch)}?recursive=1")
            self._tree = {
                item["path"]: item
                for item in data.get("tree", [])
                if isinstance(item, dict) and isinstance(item.get("path"), str)
            }
        return self._tree

    def refresh_tree(self) -> None:
        self._tree = None

    def blob_bytes(self, sha: str) -> bytes:
        data = self.request("GET", f"/repos/{self.owner}/{self.repo}/git/blobs/{sha}")
        content = str(data.get("content") or "")
        return base64.b64decode(content)

    def commits_for_path(self, repo_path: str) -> list[dict[str, str]]:
        repo_path = normalize_repo_path(repo_path)
        query = urllib.parse.urlencode({"sha": self.branch, "path": repo_path, "per_page": "10"})
        data = self.request("GET", f"/repos/{self.owner}/{self.repo}/commits?{query}")
        commits = []
        if not isinstance(data, list):
            return commits
        for item in data:
            commit = item.get("commit", {}) if isinstance(item, dict) else {}
            author = commit.get("author", {}) if isinstance(commit, dict) else {}
            commits.append(
                {
                    "sha": str(item.get("sha", ""))[:7],
                    "date": str(author.get("date", ""))[:10],
                    "author": str(author.get("name", "")),
                    "subject": str(commit.get("message", "")).splitlines()[0],
                }
            )
        return commits

    def request(self, method: str, path: str, body: dict[str, Any] | None = None) -> Any:
        if not self.token:
            raise GitHubError("GITHUB_TOKEN is not configured")
        url = f"https://api.github.com{path}"
        data = None
        headers = {
            "accept": "application/vnd.github+json",
            "authorization": f"Bearer {self.token}",
            "user-agent": "dataops-lambda",
            "x-github-api-version": "2022-11-28",
        }
        if body is not None:
            data = json.dumps(body).encode("utf-8")
            headers["content-type"] = "application/json"
        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                raw = resp.read()
        except urllib.error.HTTPError as exc:
            detail = exc.read().decode("utf-8", "replace")
            raise GitHubError(f"GitHub {method} {path} failed: HTTP {exc.code}: {detail}") from exc
        if not raw:
            return {}
        return json.loads(raw.decode("utf-8"))

    def _write_repo_file(self, repo_path: str, content: bytes) -> None:
        local = self.local_path(repo_path)
        local.parent.mkdir(parents=True, exist_ok=True)
        local.write_bytes(content)


def normalize_repo_path(path: str) -> str:
    clean = path.strip().replace("\\", "/").lstrip("/")
    if not clean or ".." in clean.split("/"):
        raise ValueError("Invalid repository path")
    return clean


def should_hydrate_tarball_path(path: str) -> bool:
    try:
        repo_path = normalize_repo_path(path)
    except ValueError:
        return False
    if not repo_path.startswith("content/"):
        return False
    if repo_path.endswith(".md"):
        return True
    return repo_path.startswith("content/images/") and Path(repo_path).suffix.lower() in CONTENT_IMAGE_EXTENSIONS


def quote_path(path: str) -> str:
    return urllib.parse.quote(path, safe="/")


_SECRET_CACHE: dict[str, str] = {}
_GITHUB_TOKEN_CACHE: str | None = None


def github_token() -> str:
    global _GITHUB_TOKEN_CACHE
    if _GITHUB_TOKEN_CACHE is not None:
        return _GITHUB_TOKEN_CACHE

    token = os.environ.get("GITHUB_TOKEN", "")
    secret_name = os.environ.get("GITHUB_TOKEN_SECRET_NAME", "")
    if token or not secret_name:
        _GITHUB_TOKEN_CACHE = token
        return token

    secret = secret_string(secret_name)
    _GITHUB_TOKEN_CACHE = secret
    return secret


def secret_string(secret_name: str) -> str:
    if secret_name in _SECRET_CACHE:
        return _SECRET_CACHE[secret_name]

    try:
        import boto3  # type: ignore[import-not-found]

        data = boto3.client("secretsmanager").get_secret_value(SecretId=secret_name)
    except Exception as exc:
        raise GitHubError(f"Could not load secret {secret_name}: {exc}") from exc

    secret = data.get("SecretString") or ""
    if not secret and data.get("SecretBinary"):
        secret = base64.b64decode(data["SecretBinary"]).decode("utf-8")
    _SECRET_CACHE[secret_name] = secret
    return secret

File: backend/src/docs/githubStore.ts
/**
 * GitHub content store.
 *
 * Ports `lambda-functions/src/lambda_functions/github_store.py` to TypeScript.
 * GitHub markdown remains the source of truth for content
 * (`_docs/TARGET_ARCHITECTURE.md`); this store reads the repo tree/files through
 * the GitHub Contents/Git Data API, keeps a `/tmp` working copy as a cache, and
 * commits on save/delete through the Contents API.
 *
 * Difference from the Python original: hydration uses the recursive Git Trees
 * API plus per-blob fetches (cached to `/tmp`) instead of a single tarball
 * download, so the store has no third-party tar dependency. The cache location,
 * source-of-truth semantics, and commit-on-save behavior are unchanged. (A
 * batch tarball download is a possible future perf optimization.)
 */

import { Buffer } from 'node:buffer';
import { mkdirSync, readFileSync, rmSync, writeFileSync, existsSync, statSync } from 'node:fs';
import { dirname, resolve, sep } from 'node:path';

import { GetSecretValueCommand, SecretsManagerClient } from '@aws-sdk/client-secrets-manager';

/** Image extensions hydrated from `content/images/` into the cache. */
export const CONTENT_IMAGE_EXTENSIONS = new Set(['.png', '.jpg', '.jpeg', '.gif', '.webp', '.svg']);

/** Raised when a GitHub API request fails. */
export class GitHubError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'GitHubError';
  }
}

/** Configuration for a {@link GithubStore}. */
export interface GithubStoreConfig {
  /** Repository owner / org, e.g. `DataTalksClub`. */
  owner: string;
  /** Repository name, e.g. `dataops`. */
  repo: string;
  /** Branch to read/commit against. Defaults to the repo default branch. */
  branch?: string;
  /** GitHub token; resolved from Secrets Manager / env by the implementation. */
  token?: string;
  /** Name of the Secrets Manager secret holding the token (lazy resolution). */
  tokenSecretName?: string;
  /** Local cache directory for the markdown working copy. Defaults to `/tmp`. */
  cacheDir?: string;
  /** Injectable `fetch` (defaults to global `fetch`); used to mock GitHub in tests. */
  fetchImpl?: typeof fetch;
}

/** One entry in the repository tree. */
export interface GithubTreeEntry {
  /** Repo-relative path, e.g. `content/sops/foo.md`. */
  path: string;
  /** Git blob/tree SHA. */
  sha: string;
  /** Entry kind. */
  type: 'blob' | 'tree';
  /** Blob size in bytes, when known. */
  size?: number;
}

/** A commit touching a given path, used for history views. */
export interface GithubCommitInfo {
  sha: string;
  message: string;
  author: string;
  /** ISO-8601 (or `YYYY-MM-DD`) commit date. */
  date: string;
}

/**
 * Read/write access to the content repository.
 *
 * Reads are served from a `/tmp` cache hydrated from GitHub; writes commit
 * directly through the GitHub Contents API and refresh the cache.
 */
export interface GithubStore {
  /** Map of repo-relative path -> tree entry. Served from the `/tmp` cache. */
  tree(): Promise<Record<string, GithubTreeEntry>>;

  /** Force-refresh the cached tree from GitHub. */
  refreshTree(): Promise<void>;

  /** Hydrate / sync the markdown working copy into the `/tmp` cache. */
  sync(): Promise<void>;

  /** Read a file's UTF-8 text by repo path. */
  readFile(repoPath: string): Promise<string>;

  /** Read a file's raw bytes by repo path. */
  readBytes(repoPath: string): Promise<Uint8Array>;

  /** Commit-on-save: create/update a file via the Contents API. */
  writeFile(repoPath: string, content: string | Uint8Array, message: string): Promise<void>;

  /** Commit-on-delete: remove a file via the Contents API. */
  deleteFile(repoPath: string, message: string): Promise<void>;

  /** Current blob SHA for a path (needed for Contents API updates). */
  currentSha(repoPath: string): Promise<string>;

  /** Commit history for a path. */
  commitsForPath(repoPath: string): Promise<GithubCommitInfo[]>;

  /** Clear any in-memory/`/tmp` cache state. */
  reset(): void;
}

const DEFAULT_CACHE_DIR = '/tmp/dataops';
const GITHUB_API = 'https://api.github.com';

/** Normalize a repo-relative path; reject traversal and absolute paths. */
export function normalizeRepoPath(path: string): string {
  const clean = path.trim().replace(/\\/g, '/').replace(/^\/+/, '');
  if (!clean || clean.split('/').includes('..')) {
    throw new Error('Invalid repository path');
  }
  return clean;
}

/** Percent-encode a path while keeping `/` separators (Python `quote_path`). */
export function quotePath(path: string): string {
  return path
    .split('/')
    .map((segment) => encodeURIComponent(segment))
    .join('/');
}

/** True when a tree/tarball path should be hydrated into the cache. */
export function shouldHydratePath(path: string): boolean {
  let repoPath: string;
  try {
    repoPath = normalizeRepoPath(path);
  } catch {
    return false;
  }
  if (!repoPath.startsWith('content/')) return false;
  if (repoPath.endsWith('.md')) return true;
  if (!repoPath.startsWith('content/images/')) return false;
  const dot = repoPath.lastIndexOf('.');
  const ext = dot >= 0 ? repoPath.slice(dot).toLowerCase() : '';
  return CONTENT_IMAGE_EXTENSIONS.has(ext);
}

/**
 * GitHub content store backed by the Contents / Git Data API with a `/tmp`
 * working-copy cache. Implements {@link GithubStore} and exposes a few extra
 * filesystem helpers used by the content API (cache paths + commit helpers).
 */
export class ContentsApiGithubStore implements GithubStore {
  readonly owner: string;
  readonly repo: string;
  readonly branch: string;
  readonly root: string;
  readonly contentRoot: string;

  private readonly fetchImpl: typeof fetch;
  private readonly configToken?: string;
  private readonly tokenSecretName?: string;
  private resolvedToken: string | null = null;
  private treeCache: Record<string, GithubTreeEntry> | null = null;
  private hydrated = false;
  private secretsClient: SecretsManagerClient | null = null;

  constructor(config: GithubStoreConfig) {
    this.owner = config.owner;
    this.repo = config.repo;
    this.branch = config.branch || 'main';
    this.root = resolve(config.cacheDir || DEFAULT_CACHE_DIR);
    this.contentRoot = resolve(this.root, 'content');
    this.fetchImpl = config.fetchImpl || fetch;
    this.configToken = config.token;
    this.tokenSecretName = config.tokenSecretName;
  }

  get githubUrl(): string {
    return `https://github.com/${this.owner}/${this.repo}`;
  }

  reset(): void {
    this.treeCache = null;
    this.hydrated = false;
    if (existsSync(this.root)) {
      rmSync(this.root, { recursive: true, force: true });
    }
  }

  /** Map a repo path to an absolute path inside the cache, guarding traversal. */
  localPath(repoPath: string): string {
    const clean = normalizeRepoPath(repoPath);
    const target = resolve(this.root, clean);
    if (target !== this.root && !target.startsWith(this.root + sep)) {
      throw new Error('Path escapes GitHub cache root');
    }
    return target;
  }

  async sync(): Promise<void> {
    if (this.hydrated) return;
    // Offline/local dev: content is already present under contentRoot (e.g. the
    // repo's own content/), so skip GitHub hydration entirely.
    if (process.env.DTC_OFFLINE === '1') {
      this.hydrated = true;
      return;
    }
    mkdirSync(this.root, { recursive: true });
    mkdirSync(this.contentRoot, { recursive: true });
    const tree = await this.tree();
    for (const entry of Object.values(tree)) {
      if (entry.type !== 'blob' || !shouldHydratePath(entry.path)) continue;
      if (existsSync(this.localPath(entry.path))) continue;
      this.writeRepoFile(entry.path, await this.blobBytes(entry.sha));
    }
    this.hydrated = true;
  }

  /** Ensure a single file is present in the cache, hydrating it if needed. */
  async ensureFile(repoPath: string): Promise<string> {
    const clean = normalizeRepoPath(repoPath);
    const local = this.localPath(clean);
    if (existsSync(local)) return local;
    // Offline/local dev: never reach for GitHub — a missing file is just ENOENT.
    if (process.env.DTC_OFFLINE === '1') {
      const err = new Error(clean) as NodeJS.ErrnoException;
      err.code = 'ENOENT';
      throw err;
    }
    const entry = (await this.tree())[clean];
    if (!entry || entry.type !== 'blob') {
      const err = new Error(clean) as NodeJS.ErrnoException;
      err.code = 'ENOENT';
      throw err;
    }
    this.writeRepoFile(clean, await this.blobBytes(entry.sha));
    return local;
  }

  async readFile(repoPath: string): Promise<string> {
    const local = await this.ensureFile(repoPath);
    return readFileSync(local, 'utf-8');
  }

  async readBytes(repoPath: string): Promise<Uint8Array> {
    const local = await this.ensureFile(repoPath);
    return new Uint8Array(readFileSync(local));
  }

  async writeFile(repoPath: string, content: string | Uint8Array, message: string): Promise<void> {
    const clean = normalizeRepoPath(repoPath);
    const bytes = typeof content === 'string' ? Buffer.from(content, 'utf-8') : Buffer.from(content);
    this.writeRepoFile(clean, bytes);
    await this.commitLocalFile(clean, message);
  }

  async deleteFile(repoPath: string, message: string): Promise<void> {
    const clean = normalizeRepoPath(repoPath);
    const local = this.localPath(clean);
    if (existsSync(local)) rmSync(local, { force: true });
    await this.deleteRepoFile(clean, message);
  }

  /** Commit the current cached bytes of a path through the Contents API. */
  async commitLocalFile(repoPath: string, message: string): Promise<void> {
    const clean = normalizeRepoPath(repoPath);
    const content = readFileSync(this.localPath(clean));
    const currentSha = await this.currentSha(clean);
    const body: Record<string, unknown> = {
      message,
      content: content.toString('base64'),
      branch: this.branch,
    };
    if (currentSha) body.sha = currentSha;
    await this.request('PUT', `/repos/${this.owner}/${this.repo}/contents/${quotePath(clean)}`, body);
    await this.refreshTree();
  }

  /** Delete a path from the repo through the Contents API. */
  async deleteRepoFile(repoPath: string, message: string): Promise<void> {
    const clean = normalizeRepoPath(repoPath);
    const currentSha = await this.currentSha(clean);
    if (!currentSha) return;
    await this.request('DELETE', `/repos/${this.owner}/${this.repo}/contents/${quotePath(clean)}`, {
      message,
      sha: currentSha,
      branch: this.branch,
    });
    await this.refreshTree();
  }

  async currentSha(repoPath: string): Promise<string> {
    const entry = (await this.tree())[normalizeRepoPath(repoPath)];
    if (!entry || entry.type !== 'blob') return '';
    return entry.sha || '';
  }

  async tree(): Promise<Record<string, GithubTreeEntry>> {
    if (this.treeCache === null) {
      const data = await this.request(
        'GET',
        `/repos/${this.owner}/${this.repo}/git/trees/${quotePath(this.branch)}?recursive=1`,
      );
      const tree: Record<string, GithubTreeEntry> = {};
      const items = Array.isArray((data as { tree?: unknown }).tree) ? (data as { tree: unknown[] }).tree : [];
      for (const raw of items) {
        if (!raw || typeof raw !== 'object') continue;
        const item = raw as Record<string, unknown>;
        if (typeof item.path !== 'string') continue;
        tree[item.path] = {
          path: item.path,
          sha: String(item.sha || ''),
          type: item.type === 'tree' ? 'tree' : 'blob',
          size: typeof item.size === 'number' ? item.size : undefined,
        };
      }
      this.treeCache = tree;
    }
    return this.treeCache;
  }

  async refreshTree(): Promise<void> {
    this.treeCache = null;
  }

  async blobBytes(sha: string): Promise<Buffer> {
    const data = await this.request('GET', `/repos/${this.owner}/${this.repo}/git/blobs/${sha}`);
    const content = String((data as { content?: unknown }).content || '');
    return Buffer.from(content, 'base64');
  }

  async commitsForPath(repoPath: string): Promise<GithubCommitInfo[]> {
    const clean = normalizeRepoPath(repoPath);
    const query = new URLSearchParams({ sha: this.branch, path: clean, per_page: '10' });
    const data = await this.request('GET', `/repos/${this.owner}/${this.repo}/commits?${query.toString()}`);
    if (!Array.isArray(data)) return [];
    return data.map((raw): GithubCommitInfo => {
      const item = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
      const commit = (item.commit && typeof item.commit === 'object' ? item.commit : {}) as Record<string, unknown>;
      const author = (commit.author && typeof commit.author === 'object' ? commit.author : {}) as Record<string, unknown>;
      return {
        sha: String(item.sha || '').slice(0, 7),
        date: String(author.date || '').slice(0, 10),
        author: String(author.name || ''),
        message: String(commit.message || '').split('\n')[0],
      };
    });
  }

  /** Issue a GitHub API request, returning the parsed JSON (or `{}`). */
  async request(method: string, path: string, body?: Record<string, unknown>): Promise<unknown> {
    const token = await this.token();
    if (!token) throw new GitHubError('GITHUB_TOKEN is not configured');
    const headers: Record<string, string> = {
      accept: 'application/vnd.github+json',
      authorization: `Bearer ${token}`,
      'user-agent': 'dataops-backend',
      'x-github-api-version': '2022-11-28',
    };
    let payload: string | undefined;
    if (body !== undefined) {
      payload = JSON.stringify(body);
      headers['content-type'] = 'application/json';
    }
    let resp: Response;
    try {
      resp = await this.fetchImpl(`${GITHUB_API}${path}`, { method, headers, body: payload });
    } catch (err) {
      throw new GitHubError(`GitHub ${method} ${path} failed: ${(err as Error).message}`);
    }
    const raw = await resp.text();
    if (!resp.ok) {
      throw new GitHubError(`GitHub ${method} ${path} failed: HTTP ${resp.status}: ${raw}`);
    }
    if (!raw) return {};
    return JSON.parse(raw);
  }

  private writeRepoFile(repoPath: string, content: Buffer): void {
    const local = this.localPath(repoPath);
    mkdirSync(dirname(local), { recursive: true });
    writeFileSync(local, content);
  }

  /** Resolve the GitHub token from config, env, or Secrets Manager (cached). */
  private async token(): Promise<string> {
    if (this.resolvedToken !== null) return this.resolvedToken;
    const configured = this.configToken || process.env.GITHUB_TOKEN || '';
    const secretName = this.tokenSecretName || process.env.GITHUB_TOKEN_SECRET_NAME || '';
    if (configured || !secretName) {
      this.resolvedToken = configured;
      return configured;
    }
    this.secretsClient ||= new SecretsManagerClient({});
    try {
      const result = await this.secretsClient.send(new GetSecretValueCommand({ SecretId: secretName }));
      const secret =
        result.SecretString ||
        (result.SecretBinary ? Buffer.from(result.SecretBinary).toString('utf-8') : '');
      this.resolvedToken = secret;
      return secret;
    } catch (err) {
      throw new GitHubError(`Could not load secret ${secretName}: ${(err as Error).message}`);
    }
  }

  /** Last-modified time (epoch seconds) of a cached file; 0 if absent. */
  updatedAt(repoPath: string): number {
    const local = this.localPath(repoPath);
    if (!existsSync(local)) return 0;
    return Math.floor(statSync(local).mtimeMs / 1000);
  }
}

/** Build a {@link GithubStoreConfig} from environment variables. */
export function githubStoreConfigFromEnv(overrides: Partial<GithubStoreConfig> = {}): GithubStoreConfig {
  return {
    owner: process.env.GITHUB_OWNER || 'DataTalksClub',
    repo: process.env.GITHUB_REPO || 'dataops',
    branch: process.env.GITHUB_BRANCH || 'main',
    cacheDir: process.env.DTC_CACHE_ROOT || DEFAULT_CACHE_DIR,
    ...overrides,
  };
}

/** Construct a GitHub content store. */
export function createGithubStore(config: GithubStoreConfig): ContentsApiGithubStore {
  return new ContentsApiGithubStore(config);
}
